core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableCorrelate.java

```diff
@@ -100,8 +100,8 @@ public Result implement(EnumerableRelImplementor implementor,
 
     Expression selector =
         EnumUtils.joinSelector(
-            joinType.returnsJustFirstInput() ? joinType.toJoinType()
-                : JoinRelType.INNER, physType,
+            joinType.returnsJustFirstInput() ? JoinRelType.INNER
+                : joinType.toJoinType(), physType,
```

Comment:
Just curious about the decision logic, here is the `returnsJustFirstInput` definition:
```java
public boolean returnsJustFirstInput() {
    switch (this) {
    case INNER:
    case LEFT:
      return false;
    case SEMI:
    case ANTI:
      return true;
    }
    throw new IllegalStateException(
        "Unable to convert " + this + " to JoinRelType");
  }
```
So, why we get a `JoinRelType.INNER` when `joinType.returnsJustFirstInput` returns true ?

Reply:
The method `EnumUtils.joinSelector(...)` requires a `JoinRelType joinType` as first parameter. Inside this method, it will use this `joinType` to verify whether the join can generate null on the left and/or on right side, in order to generate the appropriate expression each time.
In the context of `EnumerableCorrelate#implement` method, in case of `returnsJustFirstInput` (i.e. SemiJoinType is SEMI or ANTI) we will have a join that will not generate null neither on the left nor on the right expression, and the corresponding `JoinRelType` that has this behavior will be `JoinRelType.INNER`, that is why this value is passed to `EnumUtils.joinSelector(...)`. On the other hand, if the SemiJoinType `returnsJustFirstInput()==false` (i.e. INNER or LEFT) we just pass the corresponding JoinRelType via `toJoinType()` method (i.e. `SemiJoinType.LEFT` -> `JoinRelType.LEFT` ; and `SemiJoinType.INNER` -> `JoinRelType.INNER`).

Reply:
Got your idea, but the code snippet is a little confusing, if we need just a flag to decide if the join sides are null generating, can we just pass in the flag directly? Or at lest we should add some comments here to explain why the `JoinType` and `returnsJustFirstInput` definition are inconsistent.

Reply:
I wonder if `joinType.returnsJustFirstInput` is required at all

Reply:
In the current logic (I'm just fixing the code, I was not the one who implemented it like this in the first place), I think `joinType.returnsJustFirstInput` is needed. In EnumerableCorrelate, there are four different SemiJoinType, each one of them must behave as follows when calling `EnumUtils.joinSelector(...)`
1 - SemiJoinType.LEFT --> JoinRelType.LEFT (i.e. `toJoinType()`)
2 - SemiJoinType.INNER --> JoinRelType.INNER (i.e. `toJoinType()`)
3 - SemiJoinType.SEMI --> JoinRelType.INNER
4 - SemiJoinType.ANTI --> JoinRelType.INNER
Case 1 and 2 are differentiate from 3 and 4 via `returnsJustFirstInput()`, which returns false for INNER, LEFT; and true for SEMI, ANTI. I agree that maybe the ternary opearator with `returnsJustFirstInput()` method is not the most straightforward and clear way to code this, and perhaps we are "misusing" this method.
Maybe it would be simpler if this were re-written as:
```
    JoinRelType joinRelType;
    switch (joinType) {
      case LEFT:
        joinRelType = JoinRelType.LEFT;
        break;
      default:
        joinRelType = JoinRelType.INNER;
    }
    Expression selector =
        EnumUtils.joinSelector(joinRelType, physType, 
                ImmutableList.of(leftResult.physType, rightResult.physType));
```
or
```
    JoinRelType joinRelType = joinType == SemiJoinType.LEFT ? JoinRelType.LEFT : JoinRelType.INNER;
    Expression selector =
        EnumUtils.joinSelector(joinRelType, physType,
                ImmutableList.of(leftResult.physType, rightResult.physType));
```